api/src/main/java/org/eclipse/microprofile/config/spi/ConfigValueConfigSource.java

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright (c) 2020 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package org.eclipse.microprofile.config.spi;
+
+import java.util.AbstractCollection;
+import java.util.AbstractMap;
+import java.util.AbstractSet;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Extends the original {@link ConfigSource} to expose methods that return a {@link ConfigValue}. The
+ * {@link ConfigValue} allows to retrieve additional metadata associated with the configuration resolution.
+ * <p>
+ *
+ * This is to work around the limitation from the original {@link ConfigSource}. It exposes everything as plain Strings
+ * and it is not possible to retrieve additional information associated with the Configuration. The
+ * ConfigValueConfigSource tries to make this possible.
+ * <p>
+ *
+ * @author <a href="mailto:radcortez@yahoo.com">Roberto Cortez</a>
+ */
+public interface ConfigValueConfigSource extends ConfigSource {
```

Comment:
I'm not sure if we should leave this separate or add them as default methods in `ConfigSource`.

Reply:
I was thinking we should separate it completely, introducing a `ConfigValueSource` and `ConfigValueSourceProvider` API (separating also allows us to shorten the name a little bit).  We could then deprecate `ConfigSource`; for interim compatibility, we could specify that the implementation must automatically wrap old style `ConfigSource` with `ConfigValueSource`.  Then we can eventually drop those off after an acceptable period of time, and not have to deal with lots of sticky overloads.

WDYT?

Reply:
For simplicity, I prefer to add them as default methods in ConfigSource. I don't see any conflicts. Besides, I am not a fan to depricate for naming reasons. ConfigSource is very easy to be understood.

Reply:
For simplicity, I prefer the opposite.  Keep it separate.  Adding default methods makes the compatibility story far more complex.  A separate interface keeps things simple and in cases like this it is a much better way to change behavior without compromising compatibility with old implementations, and it also avoids complexity with new implementations.